soc: intel_adsp: Refactored IPC/IDC

aborisovich opened the pull request:
Refactored common IPC driver code to use "Intel ADSP" naming so it may be used for ACE SoC in future.
Refactored existing CAVS IPC API to be common for all Intel ADSP boards. Unified ipc and idc devicetree macros and registers
for all Intel ADSP families.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

carlocaione commented on drivers/Kconfig:
```diff
@@ -147,4 +147,6 @@ source "drivers/coredump/Kconfig"
 
 source "drivers/xen/Kconfig"
 
+source "drivers/ipc/Kconfig"
```
We already have an IPC subsystem definitely we do not need a new IPC driver class for this.

nashif replied:
agree, why add a lone Kconfig in here? any folders under drivers/ is considered a subsystem that needs a defined API and drivers will need to be part of this subsystem.

aborisovich replied:
Ok, we can just check whether its enabled/disabled using single generic CONFIG_INTEL_ADSP_IPC and differentiate implementation by board/family definitions. Yes this can be removed thanks.
However I will need to leave existing CAVS_IPC symbol selecting INTEL_ADSP_IPC symbol just for the sake of CIs building SOF code that needs that symbol.

nashif commented:
This is a whole lot of changes in 1 single commit, makes review VERY difficult if not impossible.

crojewsk-intel commented:
You could start by listing the number of questions your PR is answering.
In the spirit of _process_: each commit shall be atomic - provide an answer to a single question.

aborisovich commented:
> You could start by listing the number of questions your PR is answering. In the spirit of _process_: each commit shall be atomic - provide an answer to a single question.

> Good page, nice and short: https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

Guys I am not doing it from yesterday ok? I know how it should look like.
I said that it is tedious to split it now to commits, but ok I will spend that time doing it.

Commit:
soc: intel_adsp: Refactored IPC/IDC

Changes to code:
1. Renamed CAVS_IPC API from common/include/cavs_ipc.h to
common/include/intel_adsp_ipc.h. Renamed all API functions and structs -
added "intel_adsp_" prefix.
2. Moved definitions from intel-ipc-regs.h and ace-ipc-regs.g to SOC
specific headers include/<soc_name>/adsp_ipc_regs.h.
3. Added new common intel_adsp_ipc_devtree.h header with new
macros to retrieve IPC and IDC nodes and register addresses.
Put those new macros in code replacing hardcoded values outside of
devicetree.
4. Changed documentation of IDC and renamed IDC register struct
to have common name between all intel adsp socs.
5. Removed excessive docs description on cAVS IPC protocol.

Changes to Devicetree:
1. Renamed in all CAVS boards .dtsi files content in IPC nodes:
   - "cavs_host_ipc" node labels to "adsp_ipc" labels.
   - compatible "intel,cavs-host-ipc" renamed to
     "intel,adsp-host-ipc".
2. Added (previously missing) yaml file for "intel,adsp-host-ipc"
   compatible.
3. Renamed in all CAVS boards .dtsi files content in IDC nodes:
   - "idc" node labels to "adsp_idc" labels.
   - compatible "intel,cavs-idc" renamed to "intel-adsp-idc"
4. Renamed intel,cavs_idc.yaml file to intel,adsp_idc.yaml
   so it is suitable for both CAVS and ACE SoC family.
   Moved it from ipm bindings to ipc bindings where it belongs.

Changes to Kconfig:
1. Renamed existing Kconfig option CONFIG_CAVS_IPC to
   INTEL_ADSP_IPC.
2. For renamed INTEL_ADSP_IPC addded default value based on
   status of the "adsp-ipc" and "adsp-ipc" node.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

aborisovich commented:
Split PR onto two commits:
* containing refactorings done with good description on content
* adding IPC implementation for ACE1X SoC

Removed Kconfig entries from drivers/ipc as @carlocaione and @nashif suggested.

carlocaione commented on soc/xtensa/intel_adsp/common/include/intel-adsp-ipc-regs.h:
```diff
@@ -0,0 +1,94 @@
+#endif /* CONFIG_SOC_SERIES_INTEL_ACE1X */
+};
+
+#define INTEL_ADSP_IPC_BUSY BIT(31)
+#define INTEL_ADSP_IPC_DONE BIT(31)
```
Some bit for both?

aborisovich replied:
Yes it is. It's just a rename of:
![image](https://user-images.githubusercontent.com/12482515/178689906-1ef530f7-7dbe-4bb9-b360-5c0e02852de2.png)
IT's just written to different registers:
BUSY to TDR and IDR
DONE to TDA and IDA

carlocaione commented on dts/bindings/ipc/intel,adsp-idc.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2019 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
```
Nordic or Intel?

aborisovich replied:
Intel! Thanks good catch

mbolivar-nordic reviewed (commented):
ACK on the DT portions

carlocaione commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -48,24 +54,31 @@ void z_intel_adsp_ipc_isr(const void *devarg)
-		} else {
-			regs->ida = INTEL_ADSP_IPC_DONE;
-		}
+#ifdef CONFIG_SOC_SERIES_INTEL_CAVS_V15
+		regs->idd = CAVS_IPC_IDD15_DONE;
+#else
+		regs->ida = INTEL_ADSP_IPC_DONE;
+#endif
```
I'm curious about why changing `IS_ENABLED` with `ifdef` in this file.

aborisovich replied:
I think I've already answered this question.
IS_ENABLED macro expects that both CAVS_IPC_IDD15_DONE and INTEL_ADSP_IPC_DONE are defined in compilation unit.
My refactorings changes this that when you compile ACE1X it does not have CAVS_IPC_IDD15_DONE defined, because why it would have definitions for other SoC that is not currently built.
So IS_ENABLED had to be replaced with #ifdef.

kv2019i commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -50,24 +55,31 @@ void z_intel_adsp_ipc_isr(const void *devarg)
+			regs->tda = ACE_IPC_TDA1X_DONE;
+#else
 			regs->tda = INTEL_ADSP_IPC_DONE;
+#endif
```
This could potentially be cleaned up further by defining the register and done bit contents in one place and then do all the tda/tdd/ida operations without ifdefs in the code. But there are so few, that I don't think this is needed for this initial PR. The code is understandable in current form and the ifdefs have been in use before this PR (for V1.5).

kv2019i reviewed (dismissed):
Thanks @aborisovich . Splitting the patches did help in review a lot. This seems good to me now.

teburd reviewed (dismissed):
PR Looks good

andyross reviewed (dismissed):
Looks clean.  I wouldn't cry if the first patch was further split to isolate "rename" from "refactor" operations (it touches a lot of lines and it's hard to find the important parts), but that's minor.  Also some of the preprocessor work could be unified per @kv2019i, or replaced with if (IS_ENABLED()) expressions which are a little easier to read.

nashif commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -1,162 +1,178 @@
+#elif defined(CONFIG_INTEL_ADSP_CAVS)
 #include <cavs-ipc-regs.h>
+#endif
+#include <autoconf.h>
```
why are you including this? This is already included by the build system and should not be included directly.

nashif commented on dts/xtensa/intel/intel_cavs20.dtsi:
```diff
@@ -125,8 +125,9 @@
-			compatible = "intel,cavs-idc";
+		intel_adsp_idc: idc@1200 {
+			compatible = "intel,adsp-idc";
+			label = "CAVS_IDC";
```
labels are now not required and being cleaned up in the tree, this also keeps the CAVS in the name, which is beling replaced everywhere else.

serhiy-katsyuba-intel commented on soc/xtensa/intel_adsp/ace_v1x/ace-ipc-regs.h:
```diff
@@ -15,6 +15,15 @@
+ * must be cleared (!), not set (in contrary to CAVS SoC family boards).
+ * This clears BUSY on the other side of the connection in IDR register.
+ */
+#define ACE_IPC_TDA1X_DONE BIT(0)
```
There is no such thing as bit DONE at position 0 in TDA registry according to ACE documentation. TDA has only BUSY bit at position 31 and rest is MSG field (bits 30..0). Actually the comment above is correct. So this define is just wrong and misleading.

serhiy-katsyuba-intel commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -50,24 +55,31 @@ void z_intel_adsp_ipc_isr(const void *devarg)
 		regs->tdr = INTEL_ADSP_IPC_BUSY;
 		if (done && !IS_ENABLED(CONFIG_SOC_INTEL_CAVS_V15)) {
+#ifdef CONFIG_SOC_SERIES_INTEL_ACE1X
+			regs->tda = ACE_IPC_TDA1X_DONE;
```
On ACE TDA is acknowledged by clearing bit 31 (BUSY). Here is writing value 1 to TDA which, of course, clears bit 31 and that is why it works but is misleading. It is better to just write 0.
Or, alternatively, something like ACK_TDA() macro could be defined which for ACE will write 0 into TDA with an explanation what that 0 is doing. As a bonus fewer "#ifdef CONFIG_SOC_SERIES_INTEL_ACE1X" will be needed in code.

serhiy-katsyuba-intel commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -84,12 +96,13 @@ int intel_adsp_ipc_init(const struct device *dev)
+	config->regs->idd = CAVS_IPC_IDD15_DONE;
+#else
+	config->regs->ida = INTEL_ADSP_IPC_DONE;
+	config->regs->tda = INTEL_ADSP_IPC_DONE;
```
This has no effect on ACE. To ACK TDA (to clear BUSY on host side), bit 31 must be cleared on ACE, not set. Just write 0, or use ACK_TDA() as suggested above.

serhiy-katsyuba-intel commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -98,7 +111,11 @@ void intel_adsp_ipc_complete(const struct device *dev)
 	const struct intel_adsp_ipc_config *config = dev->config;
 
+#ifdef CONFIG_SOC_SERIES_INTEL_ACE1X
+	config->regs->tda = ACE_IPC_TDA1X_DONE;
```
Again, should be replace with = 0 or ACK_TDA().

aborisovich replied:
Dropping ACE implementation in this PR. It waits too long for SOF MTL upstream.

aborisovich commented:
Push update:
* removed IPC/IDC implementation for ACE SoC from this PR. It is no use to have it here while it can not be tested.
  ACE implementation will be added after Meteorlake platform is upstreamed in SOF.
* removed labels from devicetree as @nashif suggested
* moved ipc registers back to SoC family headers

kv2019i reviewed (dismissed):
Thanks @aborisovich , looks good now. Also did a test with SOF main on a Intel cAVS2.5 platform and seems to be good (as expected, but wanted to double-check).

galak commented on drivers/ipm/ipm_cavs_idc.c:
```diff
@@ -3,9 +3,6 @@
  * SPDX-License-Identifier: Apache-2.0
  */
-
-#define DT_DRV_COMPAT intel_cavs_idc
```
How are you removing this?  How can the driver still compile?

aborisovich replied:
it builds using devicetree macros.

nashif replied:
hmm, `DT_INST_IRQN` in cavs_idc_init expects DT_DRV_COMPAT, looks like this driver is not being built in CI, otherwise it would fail.

aborisovich replied:
Ok I'll have a look at it today at later time

galak reviewed (dismissed):
how does ipm_cavs_idc.c compile if you remove DT_DRV_COMPAT?

nashif commented on dts/xtensa/intel/intel_cavs15.dtsi:
```diff
@@ -123,8 +123,8 @@
 
-		idc: idc@1200 {
-			compatible = "intel,cavs-idc";
+		intel_adsp_idc: idc@1200 {
```
the intel in `intel_adsp_idc` seems redundant, the compatible already has the vendor, so just keep it `adsp_idc`.

nashif commented on dts/xtensa/intel/intel_cavs18.dtsi:
```diff
@@ -82,8 +82,8 @@
 
-		cavs_host_ipc: cavs_host_ipc@71e00 {
-			compatible = "intel,cavs-host-ipc";
+		intel_adsp_host_ipc: cavs_host_ipc@71e00 {
```
same as above and other DTS nodes, just drop the `intel_`

nashif commented on soc/xtensa/intel_adsp/common/include/cavs-ipc-regs.h:
```diff
@@ -43,23 +45,28 @@
  * interrupt would remain active.
  */
-struct cavs_ipc {
+#if defined(CONFIG_INTEL_ADSP_CAVS)
```
if this header is CAVS only, it should not be in common/include. Also, instead of having the structure ifdefed this way (v15 vs the rest) it would be nicer and cleaner to just have a header per platform and put it in `cavs/include/intel_<platform>_adsp/` instead where the `intel_adsp_ipc` has the members for the relevant platform only.

nashif commented on soc/xtensa/intel_adsp/ace_v1x/include/ace-ipc-regs.h:
```diff
@@ -4,15 +4,23 @@
+ * @brief IPC registers layout for Intel ADSP ACE1X SoC family.
+ */
+struct intel_adsp_ipc {
+#ifdef CONFIG_BOARD_INTEL_ADSP_ACE15_MTPM
```
do we need the ifdef here? who else would want this? This is also limiting, and in violation of layering, BOARD definition should not be used in SOC code. Just take it out, probably you also want to remove the enclosing ifdef (`#ifdef CONFIG_SOC_SERIES_INTEL_ACE1X`) as well.

aborisovich replied:
Well those were left here on purpose... I think I can't answer your question in public without violation of current Intel trade secrets. Lets say that there are multiple flavors of MTL board in plans and ACE 1X is not a last generation of ACE family.
That is why this very well placed limiting is added - next developer adding some new code to ACE family will not use MTL regs by accident.

aborisovich replied:
> This is also limiting, and in violation of layering, BOARD definition should not be used in SOC code.

Maybe you are right, I just followed CAVS principle by analogy...

nashif replied:
> That is why this very well placed limiting is added - next developer adding some new code to ACE family will not use MTL regs by accident.

That should not happen, thats how we have 4 generations of CAVS now, each with their own headers and they will only be included for the right generation. Adding such conditionals is something we are trying to avoid, if we have to add them for each board/soc, that will be a real mess.

nashif commented on drivers/ipm/ipm_cavs_idc.c:
```diff
@@ -3,9 +3,6 @@
  * SPDX-License-Identifier: Apache-2.0
  */
-
-#define DT_DRV_COMPAT intel_cavs_idc
```
did you forget to re-add this one?

aborisovich replied:
Yep, readded

aborisovich commented:
Can we merge? It gets merge conflicts and needs rebases pretty fast...
